### DISCUSSION TITLE
getting_started/introduction/godot_design_philosophy

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#71] [User-contributed notes]
# getting_started/introduction/godot_design_philosophy

Now that you've gotten your feet wet, let's talk about Godot's design. Every game engine is different and fits different needs. Not only do they offer a range of features, but the design of each en...

https://docs.godotengine.org/en/stable/getting_started/introduction/godot_design_philosophy.html

**Comment by Falco-Boehnke** [comment]
Took a bit: Inheritance and Aggregation seem similar but fulfill different purposes.

Aggregation allows you to compose a scene using other complex scenes, but keeps the scenes separate. It defines a 'has-a' relationship: A human 'has-a' set of eyes, 'has-a' nose, 'has-a' ability to hear.

Inheritance allows you to derive from a scene: A farmer inherits from the 'human' scene, therefore 'has-a' set of eyes, 'has-a' nose etc. The relationship is 'is-a'. The farmer 'is-a' human. But the farmer also 'has-a' hat and 'has-a' ability to harvest crops. Changes to the human aggregate to the farmer, but changes to the farmer aren't inherited to the human

**Reply by Grimslip** [reply]
Thank you for bringing this up. I had to go and research this further for my own satisfaction. After an hour or more down a new rabbit-hole, I now have an incrementally deeper understanding of some OOP principles.
It's a long, slow process for someone who started programming over 50 years ago (when "modular programming" was the latest best practice) to bring myself up to current thinking.

**Reply by Falco-Boehnke** [reply]
Welcome back then! :)

**Reply by Grimslip** [reply]
Thank you.

**Comment by Mistiarch** [comment]
Hey, not sure how likely people are to read this, but I'm a bit unsure about inheritance as described here. 
I come from an OOP background, so I do understand the general concept of inheritance, but this guide appears to say that nodes inherit from their parent nodes in the tree, which makes sense for attributes, but it also sounds weird to me for instances of classes to inherit instances of other classes, rather than the classes themselves inheriting each other. Maybe I'm just misunderstanding, or maybe this way of inheriting exists and I've just never stumbled upon it before.

**Reply by jeanmarc** [reply]
Your understanding is correct: a derived class inherits capabilities (attributes and behavior) from its ancestors. Which means that instances of that class can be treated as instances of the ancestor classes.
Instances themselves do not inherit other instances, they "are" instances of these ancestor types.

Caveat: a class can override the ancestor's behavior with any kind of different behavior, which means that it's up to the writer of the class to properly honor the inheritance claim. Users of the class should be able to rely on the expected behavior that is advertised by its ancestors.

**Reply by stariluz** [reply]
@Mistiarch I believe it was a problem of the representation they used to show the inheritance tree. It is not the scenes tree that is explained in other parts of this manual, so, the tree in the image is currently showing the inheritance tree where both, AnimatedSprite2D and Sprite2D are found.

**Comment by hxhdrn2858** [comment]
I want to mention for anyone else who gets confused reading this:

All of the "tree" and "hierarchy" being talked about on this page is referring to the _inheritance_ hierarchy. There is also a tree structure in how you organize the nodes in a Scene, and that is unrelated. This was confusing to me because of the similar UI between the "Create New Node" menu, which I believe is what's in the screenshot on this page, and the "Scene" window.
